multi-node/aws/pkg/config/config.go

```diff
@@ -0,0 +1,326 @@
+	}
+	// minify JSON
+	var buff bytes.Buffer
+	if err := json.Compact(&buff, []byte(rendered)); err != nil {
+		return nil, err
+	}
+	return buff.Bytes(), nil
+}
+
+type Config struct {
+	Cluster
+
+	ETCDEndpoints     string
+	APIServers        string
+	SecureAPIServers  string
+	APIServerEndpoint string
+	AMI               string
+
+	// Encoded TLS assets
+	TLSConfig *CompactTLSAssets
+}
+
+func (cfg Cluster) valid() error {
+	if cfg.ExternalDNSName == "" {
+		return errors.New("externalDNSName must be set")
+	}
+	if cfg.KeyName == "" {
+		return errors.New("keyName must be set")
+	}
+	if cfg.Region == "" {
+		return errors.New("region must be set")
+	}
+	if cfg.AvailabilityZone == "" {
+		return errors.New("availabilityZone must be set")
+	}
+	if cfg.ClusterName == "" {
+		return errors.New("clusterName must be set")
+	}
+
+	_, vpcNet, err := net.ParseCIDR(cfg.VPCCIDR)
+	if err != nil {
+		return fmt.Errorf("invalid vpcCIDR: %v", err)
+	}
+
+	instancesNetIP, instancesNet, err := net.ParseCIDR(cfg.InstanceCIDR)
+	if err != nil {
+		return fmt.Errorf("invalid instanceCIDR: %v", err)
+	}
+	if !vpcNet.Contains(instancesNetIP) {
+		return fmt.Errorf("vpcCIDR (%s) does not contain instanceCIDR (%s)",
+			cfg.VPCCIDR,
+			cfg.InstanceCIDR,
+		)
+	}
+
+	controllerIPAddr := net.ParseIP(cfg.ControllerIP)
+	if controllerIPAddr == nil {
+		return fmt.Errorf("invalid controllerIP: %s", cfg.ControllerIP)
+	}
+	if !instancesNet.Contains(controllerIPAddr) {
+		return fmt.Errorf("instanceCIDR (%s) does not contain controllerIP (%s)",
+			cfg.InstanceCIDR,
+			cfg.ControllerIP,
+		)
+	}
+
+	podNetIP, podNet, err := net.ParseCIDR(cfg.PodCIDR)
+	if err != nil {
+		return fmt.Errorf("invalid podCIDR: %v", err)
+	}
+	if vpcNet.Contains(podNetIP) {
+		return fmt.Errorf("vpcCIDR (%s) overlaps with podCIDR (%s)", cfg.VPCCIDR, cfg.PodCIDR)
+	}
+
+	serviceNetIP, serviceNet, err := net.ParseCIDR(cfg.ServiceCIDR)
+	if err != nil {
+		return fmt.Errorf("invalid serviceCIDR: %v", err)
+	}
+	if vpcNet.Contains(serviceNetIP) {
+		return fmt.Errorf("vpcCIDR (%s) overlaps with serviceCIDR (%s)", cfg.VPCCIDR, cfg.ServiceCIDR)
+	}
+	if podNet.Contains(serviceNetIP) || serviceNet.Contains(podNetIP) {
+		return fmt.Errorf("serviceCIDR (%s) overlaps with podCIDR (%s)", cfg.ServiceCIDR, cfg.PodCIDR)
+	}
+
+	kubernetesServiceIPAddr := net.ParseIP(cfg.KubernetesServiceIP)
+	if kubernetesServiceIPAddr == nil {
+		return fmt.Errorf("Invalid kubernetesServiceIP: %s", cfg.KubernetesServiceIP)
+	}
+	if !serviceNet.Contains(kubernetesServiceIPAddr) {
+		return fmt.Errorf("serviceCIDR (%s) does not contain kubernetesServiceIP (%s)", cfg.ServiceCIDR, cfg.KubernetesServiceIP)
+	}
+
+	dnsServiceIPAddr := net.ParseIP(cfg.DNSServiceIP)
+	if dnsServiceIPAddr == nil {
+		return fmt.Errorf("Invalid dnsServiceIP: %s", cfg.DNSServiceIP)
+	}
+	if !serviceNet.Contains(dnsServiceIPAddr) {
+		return fmt.Errorf("serviceCIDR (%s) does not contain dnsServiceIP (%s)", cfg.ServiceCIDR, cfg.DNSServiceIP)
+	}
```

Comment:
Also should add a test that dnsServiceIPAddr != first ip in serviceCIDR and != kubernetesServiceIPAddr